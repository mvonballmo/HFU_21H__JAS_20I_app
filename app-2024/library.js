"use strict";

let _counter = 1;
let m = 2;
show(_counter + m);

function counter()
{
  return _counter;
}

function increment()
{
  return ++_counter;
}

function decrement()
{
  return --_counter;
}

function show(value)
{
  console.log(value);
}

// TODO Re-export methods with ES6 Modules
// https://trello.com/c/6dpjfKIP/6-fix-js-module-exports-so-it-works-in-the-browser

module.exports = {
  counter,
  increment,
  decrement
}