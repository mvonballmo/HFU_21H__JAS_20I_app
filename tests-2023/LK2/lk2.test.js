/**
 * @jest-environment jsdom
 */

import { beforeEach, describe, expect, test } from "@jest/globals";
import "isomorphic-fetch";
import {} from "./lk2.library";

describe("LK2", () => {
  describe("DOM", () => {
    test("get single element by ID", () => {
      document.body.innerHTML = `
      <div>
        <button id="a" class="primary hl"><span class="strong">Test</span></button>
        <button id="b" class="secondary">Run</button>
      </div>
    `;

      const button = document;

      expect(button.id).toBe("a");
      expect(button.tagName).toBe("BUTTON");
      expect(button.className).toBe("primary hl");
      expect([...button.classList]).toEqual(["primary", "hl"]);
      expect(button.innerHTML).toBe(`<span class="strong">Test</span>`);
      expect(button.outerHTML).toBe(`<button id="a" class="primary hl"><span class="strong">Test</span></button>`);
    });
  });

  describe("Events", () => {
    test("only button should get event", () => {
      document.body.innerHTML = `
      <div>
        <button/>
      </div>
    `;

      const [div] = document.getElementsByTagName("div");
      const button = div.firstElementChild;

      let clicks = [];

      function handleClick(e) {
        clicks.push(e.currentTarget.nodeName);
        e.stopPropagation();
      }

      document.addEventListener("click", handleClick);
      try {
        div.addEventListener("click", handleClick);
        button.addEventListener("click", handleClick);

        button.click();

        expect(clicks).toEqual(["#document"]);
      } finally {
        document.removeEventListener("click", handleClick);
      }
    });

    test("only document should get event", () => {
      document.body.innerHTML = `
      <div>
        <button/>
      </div>
    `;

      const [div] = document.getElementsByTagName("div");
      const button = div.firstElementChild;

      let clicks = [];

      function handleClick(e) {
        clicks.push(e.currentTarget.nodeName);
      }

      document.addEventListener("click", handleClick, { capture: false });
      try {
        div.addEventListener("click", handleClick);
        button.addEventListener("click", handleClick);

        button.click();

        expect(clicks).toEqual(["#document"]);
      } finally {
        // It doesn't matter if document event-listeners in other tests are still attached,
        // but this one captures all events and then stops propagation, which breaks the
        // expectation in the next text (for preventDefault())
        document.removeEventListener("click", handleClick, { capture: false });
      }
    });
  });

  describe("Forms and Inputs", () => {
    test("Find named element in a form", () => {
      document.body.innerHTML = `
      <form id="vehicles">
        <input type="text">
        <input id="lastName" type="text">
        <select id="numCars">
          <option id="one">One</option>
          <option id="two">Two</option>
          <option id="three">Three</option>
        </select>
      </form>
    `;

      let firstInput = document.getElementsByTagName("form")[0].getElementsByTagName("input")[0];

      expect(firstInput.id).toBe("firstName");
      expect(firstInput.disabled).toBeTruthy();
    });
  });

  describe("Web components", () => {
    test("initialize", async () => {
      document.body.innerHTML = `
      <my-element></my-element>
    `;

      const [myElement] = document.getElementsByTagName("my-element");

      expect(myElement.innerHTML).toMatchSnapshot();
    });

    test("set attribute", async () => {
      document.body.innerHTML = `
      <my-element></my-element>
    `;

      const [myElement] = document.getElementsByTagName("my-element");

      myElement.setAttribute("title", "This is my excellent title.");

      expect(myElement.innerHTML).toMatchSnapshot();
    });
  });

  describe("Cookies", () => {
    function clearCookies() {
      const cookies = document.cookie.split(";");
      const expiryDate = new Date(0).toUTCString();

      for (let i = 0; i < cookies.length; i++) {
        const cookie = cookies[i];
        const [key] = cookie.trim().split("=");
        document.cookie = `${key}=; expires=${expiryDate}`;
      }
    }

    function getCookieMap() {
      const cookie = document.cookie.split(";").map(s => {
        const [key, value] = s.trim().split("=");
        return { key, value };
      });
      return cookie;
    }

    function getCookieObject() {
      const cookie = {};

      for (const pair of document.cookie.split(";")) {
        const [key, value] = pair.trim().split("=");
        cookie[key] = value;
      }
      return cookie;
    }

    beforeEach(() => {
      clearCookies();
    });

    test("get key/value pairs", () => {
      document.cookie = "something=John";
      document.cookie = "expertMode=1";

      expect(getCookieMap()).toMatchSnapshot();
    });

    test("get values in object", () => {
      document.cookie = "something=Mark";
      document.cookie = "expertMode=John";

      expect(getCookieObject()).toMatchSnapshot();
    });

    test("store and load JSON", () => {
      const person = {
        first: "Bob",
        last: "Hoffman",
        age: 34,
        company: {
          name: "Apple",
        },
      };

      document.cookie = "???";

      expect(document.cookie).toMatchSnapshot();

      const loadedPerson = JSON.parse(getCookieObject()["person"]);

      expect(loadedPerson.first).toBe("Bob");
      expect(loadedPerson.last).toBe("Hoffman");
      expect(loadedPerson.age).toBe(34);
      expect(loadedPerson.company.name).toBe("Apple");
    });
  });

  describe("Storage", () => {
    beforeEach(() => {
      localStorage.clear();
    });

    test("get key/value pairs", () => {
      function* getLocalStorageItems() {
        for (let i = 0; i < localStorage.length; i++) {
          const key = localStorage.key(i);
          yield { key, value: localStorage.getItem(key) };
        }
      }

      localStorage.setItem("lastUsername", "Bob");

      const local = [...getLocalStorageItems()];

      expect(local).toMatchSnapshot();
    });

    test("get values in object", () => {
      function getLocalStorageAsObject() {
        const local = {};

        for (let i = 0; i < localStorage.length; i++) {
          const key = localStorage.key(i);
          local[key] = localStorage.getItem(key);
        }

        return local;
      }

      // Hint: Does this test even set anything in the localStorage yet?

      expect(getLocalStorageAsObject()).toMatchSnapshot();
    });
  });

  describe("Timers and intervals", () => {
    test("set interval of 1/10 second", done => {
      let counter = 0;

      const t = setInterval(() => {
        counter += 1;

        if (counter > 3) {
          // what do you do to stop the interval?
        }
      }, 100);

      setTimeout(() => {
        expect(counter).toBe(4);

        done();
      }, 700);
    });
  });
});
