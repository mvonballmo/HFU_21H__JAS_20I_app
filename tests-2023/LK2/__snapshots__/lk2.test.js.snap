// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LK2 Cookies get key/value pairs 1`] = `
Array [
  Object {
    "key": "lastUsername",
    "value": "John",
  },
  Object {
    "key": "expertMode",
    "value": "true",
  },
]
`;

exports[`LK2 Cookies get values in object 1`] = `
Object {
  "expertMode": "true",
  "lastUsername": "John",
}
`;

exports[`LK2 Cookies store and load JSON 1`] = `"person={\\"first\\":\\"Bob\\",\\"last\\":\\"Hoffman\\",\\"age\\":34,\\"company\\":{\\"name\\":\\"Apple\\"}}"`;

exports[`LK2 Storage get key/value pairs 1`] = `
Array [
  Object {
    "key": "lastUsername",
    "value": "John",
  },
  Object {
    "key": "expertMode",
    "value": "true",
  },
]
`;

exports[`LK2 Storage get values in object 1`] = `
Object {
  "expertMode": "true",
  "lastUsername": "John",
}
`;

exports[`LK2 Web components initialize 1`] = `"<b>Title</b>"`;

exports[`LK2 Web components set attribute 1`] = `"<b>This is my excellent title.</b>"`;
